resolver.go

```diff
@@ -34,6 +35,63 @@ type Resolver struct {
 	cache   *cache
 	expire  bool
 	timeout time.Duration
+	dialer  *net.Dialer
+}
+
+// ResolverOption configures a single resolver setting
+type ResolverOption resolverOption
+
+// resolverOption is the private resolver option type
+type resolverOption func(r *Resolver) error
+
+// NewResolver creates a resolver with the given cache capacity and a set of extra options.
+func NewResolver(options ...ResolverOption) (*Resolver, error) {
+	r := &Resolver{
+		cache:   newCache(0, false),
+		expire:  false,
+		timeout: Timeout,
+	}
+	for _, option := range options {
+		if err := option(r); err != nil {
+			return nil, err
+		}
+	}
+	return r, nil
+}
```

Comment:
- Can you move this to next to the other constructors for `Resolver`?
- Can you change the implementation of the existing constructors to use `NewResolver` instead of creating a `Resolver` directly?
- Can you mark the legacy constructors as deprecated?

Reply:
Done.

I grepped the golang sources and they seem to use a very terse `// Deprecated: Use XXX instead.` most of the time. I used the same style here.

The returned error is ignored in the legacy constructors. This is safe right now, keeps source compatibility and seems to be in line with a slow deprecation. Let me know if you prefer it in a different way!